core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala

```diff
@@ -981,11 +983,20 @@ private[spark] class DAGScheduler(
       finalStage = createResultStage(finalRDD, func, partitions, jobId, callSite)
     } catch {
       case e: BarrierJobSlotsNumberCheckFailed =>
-        logWarning(s"The job $jobId requires to run a barrier stage that requires more slots " +
-          "than the total number of slots in the cluster currently.")
         // If jobId doesn't exist in the map, Scala coverts its value null to 0: Int automatically.
         val numCheckFailures = barrierJobIdToNumTasksCheckFailures.compute(jobId,
           (_: Int, value: Int) => value + 1)
+        val retryCount = numCheckFailures - 1
+        val retryMessage = if (retryCount == 0) {
+          ""
+        } else {
+          s" (Retry ${retryCount}/$maxFailureNumTasksCheck failed)"
+        }
+
+        logWarning(s"The job $jobId requires to run a barrier stage " +
```

Comment:
I think this needs to be rewritten a little, and doesn't need the special case above
```
Barrier stage in job $jobId  requires ${...} slots, but only ${...} are available. Failure ${numCheckFailures} / ${maxFailureNumTasksCheck}
```

Reply:
My first idea was like what you suggests but If we so, we get following messages.
```
19/08/22 01:48:34 WARN DAGScheduler: Barrier stage in job 0  requires 3 slots, but only 2 are available. Failure 1 / 3
19/08/22 01:48:49 WARN DAGScheduler: Barrier stage in job 0  requires 3 slots, but only 2 are available. Failure 2 / 3
19/08/22 01:49:04 WARN DAGScheduler: Barrier stage in job 0  requires 3 slots, but only 2 are available. Failure 3 / 3
19/08/22 01:49:19 WARN DAGScheduler: Barrier stage in job 0  requires 3 slots, but only 2 are available. Failure 4 / 3
```
`Failure 4 / 3` looks weird. Another solution would using `maxFailureNumTaskCheck + 1` as the number of maximum attempt.

Reply:
How about rendering a message like "Will retry up to _x_ more times?"

Reply:
Hmm... it's like as follows?

```
19/08/22 02:54:37 WARN DAGScheduler: Barrier stage in job 0 requires 3 slots, but only 2 are available. Will retry up to 3 more times
19/08/22 02:54:52 WARN DAGScheduler: Barrier stage in job 0 requires 3 slots, but only 2 are available. Will retry up to 2 more times
19/08/22 02:55:07 WARN DAGScheduler: Barrier stage in job 0 requires 3 slots, but only 2 are available. Will retry up to 1 more times
19/08/22 02:55:22 WARN DAGScheduler: Barrier stage in job 0 requires 3 slots, but only 2 are available. Will retry up to 0 more times
org.apache.spark.scheduler.BarrierJobSlotsNumberCheckFailed: [SPARK-24819]: Barrier execution mode does not allow run a barrier stage that requires more slots than the total number of slots in the cluster currently. Please init a new cluster with more CPU cores or repartition the input RDD(s) to reduce the number of slots required to run this barrier stage.
```

Reply:
Heh OK I accept that's a little odd too, but I think it's understandable enough and a little less complex to deal with.